src/easydynamics/utils/convolution.py

```diff
@@ -0,0 +1,597 @@
+
+    Args:
+        x : np.ndarray
+            1D array of x values where the convolution is evaluated.
+        sample_model : SampleModel or ModelComponent
+            The sample model to be convolved.
+        resolution_model : SampleModel or ModelComponent
+            The resolution model to convolve with.
+        offset : Parameter, float, or None, optional
+            The offset to apply to the x values before convolution.
+        method : str, optional
+            The convolution method to use: 'analytical' or 'numerical'. Default is 'analytical'.
+        upsample_factor : int, optional
+            The factor by which to upsample the input data before numerical convolution. Default is 0 (no upsampling).
+        extension_factor : float, optional
+            The factor by which to extend the input data range before numerical convolution. Default is 0.2.
+        temperature : Parameter, float, or None, optional
+            The temperature to use for detailed balance calculations. Default is None.
+        temperature_unit : str or sc.Unit, optional
+            The unit of the temperature parameter. Default is 'K'.
+        x_unit : str or sc.Unit, optional
+            The unit of the x parameter. Default is 'meV'.
+        normalize_detailed_balance : bool, optional
+            Whether to normalize the detailed balance factor. Default is True.
+    """
+
+    # Input validation
+    if not isinstance(x, np.ndarray):
+        raise TypeError(
+            f"`x` is an instance of {type(x).__name__}, but must be a numpy array."
+        )
+
+    x = np.asarray(x, dtype=float)
+    if x.ndim != 1 or not np.all(np.isfinite(x)):
+        raise ValueError("`x` must be a 1D finite array.")
+
+    if not isinstance(sample_model, (SampleModel, ModelComponent)):
+        raise TypeError(
+            f"`sample_model` is an instance of {type(sample_model).__name__}, but must be SampleModel or ModelComponent."
+        )
+
+    if not isinstance(resolution_model, (SampleModel, ModelComponent)):
+        raise TypeError(
+            f"`resolution_model` is an instance of {type(resolution_model).__name__}, but must be SampleModel or ModelComponent."
+        )
+
+    if isinstance(sample_model, SampleModel):
+        if not sample_model.components:
+            raise ValueError("SampleModel must have at least one component.")
+
+    if isinstance(resolution_model, SampleModel):
+        if not resolution_model.components:
+            raise ValueError("ResolutionModel must have at least one component.")
+
+    # Handle offset
+    if offset is None:
+        off = 0.0
+    elif isinstance(offset, Parameter):
+        off = offset.value
+    elif isinstance(offset, Numerical):
+        off = float(offset)
+    else:
+        raise TypeError(
+            f"Expected offset to be Parameter, number, or None, got {type(offset)}"
+        )
+
+    # Handle temperature
+    T = None
```

Comment:
Avoid single letter variable names unless used in tight loops/list comprehension.